src/util/file_log.rs

```diff
@@ -0,0 +1,210 @@
+        };
+        Ok(ret)
+    }
+
+    fn open(&mut self) {
+        self.file = open_log_file(&self.file_path).unwrap()
+    }
+
+    fn should_rollover(&mut self) -> bool {
+        time::now() > self.rollover_time
+    }
+
+    fn do_rollover(&mut self) {
+        self.close();
+        let mut s = self.file_path.clone();
+        s.push_str(".");
+        s.push_str(&time::strftime("%Y%m%d", &self.rollover_time).unwrap());
+        fs::rename(&self.file_path, &s).unwrap();
+        self.update_rollover_time();
+        self.open()
+    }
+
+    fn update_rollover_time(&mut self) {
+        let now = time::now();
+        self.rollover_time = compute_rollover_time(now);
+    }
+
+    fn close(&mut self) {
+        self.file.flush().unwrap()
+    }
+}
+
+pub fn init(level: &str, file_path: &str) -> Result<(), SetLoggerError> {
+    let l = logger::get_level_by_string(level);
+    log::set_logger(|max_log_level| {
+        max_log_level.set(l);
+        Box::new(RotatingFileLogger::new(level, file_path).unwrap())
+    })
+}
+
+/// A log implemetation which writes to file and rotates by day.
+pub struct RotatingFileLogger {
+    level: LogLevelFilter,
+    core: Arc<RwLock<RotatingFileLoggerCore>>,
+}
+
+impl RotatingFileLogger {
+    pub fn new(level: &str, file_path: &str) -> io::Result<RotatingFileLogger> {
+        let core = try!(RotatingFileLoggerCore::new(file_path));
+        let ret = RotatingFileLogger {
+            level: logger::get_level_by_string(level),
+            core: Arc::new(RwLock::new(core)),
+        };
+        Ok(ret)
+    }
+
+}
+
+impl Log for RotatingFileLogger {
+    fn enabled(&self, metadata: &LogMetadata) -> bool {
+        metadata.level() <= self.level
+    }
+
+    fn log(&self, record: &LogRecord) {
+        if self.enabled(record.metadata()) {
+            let mut core = self.core.wl();
+            if core.should_rollover() {
+                core.do_rollover()
+            };
+            let now = time::now();
+            let _ = write!(core.file,
+                           "{},{:03} {}:{} - {:5} - {}\n",
+                           time::strftime("%Y-%m-%d %H:%M:%S", &now).unwrap(),
+                           now.tm_nsec / NANOSECONDS_PER_MILLISECOND,
+                           record.location().file().rsplit('/').nth(0).unwrap(),
+                           record.location().line(),
+                           record.level(),
+                           record.args());
+        }
+    }
+}
+
+impl Drop for RotatingFileLogger {
+    fn drop(&mut self) {
+        let mut core = self.core.wl();
+        core.close()
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    extern crate log;
+    extern crate rand;
+    use time::{self, Timespec};
+    use std::io::prelude::*;
+    use std::fs::OpenOptions;
+    use tempdir::TempDir;
+    use super::*;
+
+    extern crate utime;
+
+    #[test]
+    fn test_rotating_file_logger() {
+        let log_file = TempDir::new("test_rotating_file_logger.log")
```

Comment:
The `TempDir` instance should be kept, otherwise the temporary directory may be deleted immediately. The generated string should be only treated as a directory path.